src/main/java/servlets/OAuthCallbackServlet.java

```diff
@@ -57,29 +51,63 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
       return;
     }
 
-    // OAuth provider should have given us a code, otherwise something is wrong here
-    if (code == null || code.isEmpty()) {
+    // Check that the OAuth provider gave us the authorization code.
+    if (authCode == null || authCode.isEmpty()) {
       response.setStatus(400);
       return;
     }
 
-    String tokenRequestBody = String.format("%s?%s&%s&%s&%s", GRANT_TYPE, CODE,
-      REDIRECT_URI, CLIENT_ID, CLIENT_SECRET);
+    String tokenRequestBody = String.format("%s&%s&%s&%s&%s", 
+      OAuthConstants.GRANT_TYPE, OAuthConstants.AUTH_CODE + authCode, 
+      getRedirectUri(), OAuthConstants.CLIENT_ID, getClientSecret());
 
-    // Request the tokens.
+    // Request the access tokens.
     HttpClient httpClient = HttpClient.newHttpClient();
-    HttpRequest tokenRequest = HttpRequest.newBuilder(URI.create(TOKEN_URI))
-      .header("Content-Type", "application/x-www-form-urlencoded")
-      .POST(BodyPublishers.ofString(tokenRequestBody)).build();
+    HttpRequest tokenRequest = HttpRequest.newBuilder(URI.create(
+      OAuthConstants.TOKEN_URI)).header("Content-Type", 
+      "application/x-www-form-urlencoded").POST(BodyPublishers.ofString(
```

Comment:
This method would be a little easier to parse through if this was split into a few functions. Maybe move this to a "fetchAccessTokens" or "buildAccessTokensRequest" method?

Reply:
I created a buildTokenRequest() and  parseAccessToken() method. That cleaned up the doGet() a little bit.